skrifa/src/util.rs

```diff
@@ -0,0 +1,151 @@
+//! Internal helpers.
+
+use core::fmt;
+
+/// SmallVec like implementation but for fixed sized arrays whose
+/// size is only known at runtime.
+///
+/// Used to avoid allocations when possible for small or temporary
+/// arrays.
+///
+/// Consider taking on a smallvec dependency in the future.
+#[derive(Clone)]
+pub struct SmallArray<T, const N: usize>
+where
+    T: Copy,
+{
+    storage: SmallStorage<T, N>,
+}
+
+impl<T, const N: usize> SmallArray<T, N>
+where
+    T: Copy,
+{
+    pub fn new(initial_value: T, len: usize) -> Self {
+        Self {
+            storage: if len <= N {
+                SmallStorage::Inline([initial_value; N], len)
+            } else {
+                SmallStorage::Heap(vec![initial_value; len])
+            },
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn len(&self) -> usize {
+        match &self.storage {
+            SmallStorage::Inline(_, len) => *len,
+            SmallStorage::Heap(vec) => vec.len(),
+        }
+    }
+
+    pub fn as_slice(&self) -> &[T] {
+        match &self.storage {
+            SmallStorage::Inline(array, len) => &array[..*len],
+            SmallStorage::Heap(vec) => vec.as_slice(),
+        }
+    }
+
+    pub fn as_mut_slice(&mut self) -> &mut [T] {
+        match &mut self.storage {
+            SmallStorage::Inline(array, len) => &mut array[..*len],
+            SmallStorage::Heap(vec) => vec.as_mut_slice(),
+        }
+    }
+}
+
+impl<T, const N: usize> fmt::Debug for SmallArray<T, N>
+where
+    T: Copy + fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_list().entries(self.as_slice().iter()).finish()
+    }
+}
```

Comment:
The only other traits I'd consider implementing here are `AsRef`/`AsMut`, and maybe `Deref`/`DerefMut` also for completeness we could consider `Borrow`/`BorrowMut`, although those are not used that often and we can probably wait until we actually need them.


Also: If we implement `Deref` we might not need or own iter impls, if we can just inherit the ones on the slice type?

Reply:
Since this is an internal type, basically a stopgap until we can take on smallvec, I didn’t want to flesh it out with more code than is necessary to implement our required behavior. Less code = fewer bugs?

Reply:
If it's internal, can more things be `pub(crate)` instead of `pub`? I would prefer to minimize the public API as much as possible to avoid situations where we unintentionally make breaking changes.

Reply:
Done in #445! I also went ahead and added `Deref`/`DerefMut` based on your suggestion.